src/dbnode/storage/bootstrap/bootstrapper/fs/source_index_test.go

```diff
@@ -155,3 +250,139 @@ func TestBootstrapIndex(t *testing.T) {
 		}
 	}
 }
+
+func TestBootstrapIndex(t *testing.T) {
+	dir := createTempDir(t)
+	defer os.RemoveAll(dir)
+
+	times := newTestBootstrapIndexTimes(testTimesOptions{
+		numBlocks: 2,
+	})
+
+	writeTSDBGoodTaggedSeriesDataFiles(t, dir, testNs1ID, times.start)
+
+	src := newFileSystemSource(newTestOptions(dir))
+	res, err := src.ReadIndex(testNsMetadata(t), times.shardTimeRanges,
+		testDefaultRunOpts)
+	require.NoError(t, err)
+
+	indexResults := res.IndexResults()
+	validateGoodTaggedSeries(t, times.start, indexResults)
+}
+
+func TestBootstrapIndexIncremental(t *testing.T) {
+	dir := createTempDir(t)
+	defer os.RemoveAll(dir)
+
+	times := newTestBootstrapIndexTimes(testTimesOptions{
+		numBlocks: 2,
+	})
+
+	writeTSDBGoodTaggedSeriesDataFiles(t, dir, testNs1ID, times.start)
+
+	opts := newTestOptionsWithPersistManager(t, dir)
+	scope := tally.NewTestScope("", nil)
+	opts = opts.SetInstrumentOptions(opts.InstrumentOptions().SetMetricsScope(scope))
+
+	runOpts := testDefaultRunOpts.SetIncremental(true)
+
+	src := newFileSystemSource(opts).(*fileSystemSource)
+	res, err := src.ReadIndex(testNsMetadata(t), times.shardTimeRanges,
+		runOpts)
+	require.NoError(t, err)
+
+	// Check that single persisted segment got written out
+	infoFiles := fs.ReadIndexInfoFiles(src.fsopts.FilePathPrefix(), testNs1ID,
+		src.fsopts.InfoReaderBufferSize())
+	require.Equal(t, 1, len(infoFiles))
+
+	for _, infoFile := range infoFiles {
+		require.NoError(t, infoFile.Err.Error())
+		require.Equal(t, times.start.UnixNano(), infoFile.Info.BlockStart)
+		require.Equal(t, testIndexBlockSize, time.Duration(infoFile.Info.BlockSize))
+		require.Equal(t, persist.FileSetFlushType, persist.FileSetType(infoFile.Info.FileType))
+		require.Equal(t, 1, len(infoFile.Info.Segments))
+		require.Equal(t, 1, len(infoFile.Info.Shards))
+		require.Equal(t, testShard, infoFile.Info.Shards[0])
+	}
+
+	indexResults := res.IndexResults()
+
+	// Check that the segment is not a mutable segment for this block
+	block, ok := indexResults[xtime.ToUnixNano(times.start)]
+	require.True(t, ok)
+	require.Equal(t, 1, len(block.Segments()))
+	_, mutable := block.Segments()[0].(segment.MutableSegment)
+	require.False(t, mutable)
+
+	// Check that the second segment is mutable and was not written out
+	block, ok = indexResults[xtime.ToUnixNano(times.start.Add(testIndexBlockSize))]
+	require.True(t, ok)
+	require.Equal(t, 1, len(block.Segments()))
+	_, mutable = block.Segments()[0].(segment.MutableSegment)
+	require.True(t, mutable)
+
+	// Validate results
+	validateGoodTaggedSeries(t, times.start, indexResults)
+
+	// Validate that read the block and no blocks were written
+	// (ensure incremental didn't write it back out again)
+	counters := scope.Snapshot().Counters()
+	require.Equal(t, int64(0), counters["fs-bootstrapper.persist-index-blocks-read+"].Value())
```

Comment:
I see what you were talking about earlier with relying on metrics. Tests came out quite clean!